src/boot/AppEventHandlers.js

```diff
@@ -20,16 +20,36 @@ import {
 } from '../actions';
 
 /**
- * Part of the type of upstream's `NetInfo`.
- *
- * Based on docs: https://facebook.github.io/react-native/docs/netinfo
- *
- * TODO move this to a libdef, and/or get an explicit type into upstream.
+ * Part of the interface from react-native-netinfo.
+ * https://github.com/react-native-community/react-native-netinfo/tree/v3.2.1
  */
-type ConnectionInfo = {|
-  type: 'none' | 'wifi' | 'cellular' | 'unknown' | 'bluetooth' | 'ethernet' | 'wimax',
-  effectiveType: '2g' | '3g' | '4g' | 'unknown',
-|};
+// TODO: upgrade to 4.x.x so that we can use the `flow-typed` versions.
+// Requires RN 0.60+.
+type NetInfoStateType =
+  | 'unknown'
+  | 'none'
+  | 'cellular'
+  | 'wifi'
+  | 'bluetooth'
+  | 'ethernet'
+  | 'wimax'
+  | 'vpn'
+  | 'other';
+type NetInfoConnectedDetails = {
+  isConnectionExpensive: boolean,
+};
+
+type NetInfoState = {
+  /** The type of the current connection. */
+  type: NetInfoStateType,
+
+  /** If there is an active network connection. Note that this DOES NOT mean
+      that internet is reachable. */
+  isConnected: boolean,
+
+  /** Depends on the `type` value. We don't use this. */
+  details: mixed & NetInfoConnectedDetails,
```

Comment:
> Hmm interesting -- why the `mixed &`? Because `mixed` is the top of the type poset, naively it seems like `mixed & A` should be the same as `A`.

Yes, that was what I was thinking. The idea (which I failed to note in a comment) was that `mixed` was a stand-in for a complicated dependent type-fragment, which in [the original source](https://github.com/react-native-community/react-native-netinfo/blob/master/src/internal/types.ts) is passed in as a generic argument to this type.

But...

> Trying it out, it looks like with this version if we write `netInfoState.details.foo`, Flow gives no error. If it's just `NetInfoConnectionDetails`, it does.

... that's certainly not intended behavior!

> We probably want the error, right?

Yes. Yes indeed. I'll make sure we get one, and push the questionable typing into comments where it belongs.

> Or we could write `null | NetInfoConnectedDetails` ...

No, it should always be present and available.

Reply:
> Yes. Yes indeed. I'll make sure we get one, and push the questionable typing into comments where it belongs.

:+1: 

Flow does seem to need some checking up on sometimes. :-/ The `&` operator is on my rough mental list of constructs that really don't seem to wind up meaning what, as best I can tell, they're supposed to mean. For that reason I mostly avoid using it.


> > Or we could write `null | NetInfoConnectedDetails` ...
> 
> No, it should always be present and available.

I was getting the `null` from here:
https://github.com/react-native-community/react-native-netinfo/tree/v3.2.1#type-is-none-or-unknown

> #### `type` is `none` or `unknown`
>
> `details` is `null`.

Reply:
Argh. No, you're right; I misread the TypeScript source. (See `NetInfoDisconnectedState` at link above.)